### PR TITLE
Wildcards in Permissions Policy Origins

### LINKED ISSUE #479: Client Hint delegation to multiple subdomains
Hi everyone!

As part of the User Agent Reduction initiative (in Chrome), we (Akamai mPulse RUM) are implementing some changes for our product to request the information that will soon be frozen/removed from the UA string via Client Hints instead.  For example, we would like to continue being able to classify RUM visitors by their Platform Version (and not just Platform, which is included as a default CH header).

For our customers, the Akamai CDN (or their origins) will be injecting the necessary `Accept-CH` headers, e.g:

```
accept-ch: Sec-CH-UA-Platform-Version
```

Our product interacts with a few domains that we would like the Platform Version sent to, such as:

* `c.go-mpulse.net` to fetch a configuration file
* `*.akstat.io` to send the RUM beacon

So we were planning on including a `Permissions-Policy` header to delegate the requested Client Hints to those domains.  For example, we were expecting to be able to send something like:

```
permissions-policy: ch-ua-platform-version=(self "https://*.akstat.io" "https://c.go-mpulse.net") 
```

This works fine for `c.go-mpulse.net` but our request of a wildcard in `*.akstat.io` does not work.

Our `*.akstat.io` domains receiving beacons have 100s+ subdomains mapping to specific paths we want the beacon to take, i.e. it may be `abcd1234.akstat.io` for one visitor and `efgh5678.akstat.io` for another.  We don't know the domain at the time the HTML page is sent to the browser.

We were hoping that a wildcard (subdomain) match for `*.akstat.io` in the `Permissions-Policy` would work, but that does not seem to be the case (and isn't spec'd as such).

Without making any changes to the spec, we brainstormed a few options to resolve this:

1. Delegate to all subdomains: `ch-ua-platform-version=(self "*")`
    * This is not ideal, as it sends whatever Client Hints we're requesting to other 3P domains that aren't requesting it
    * This will take additional bandwidth on all other 3P requests
    * For the sake of privacy, we would like to avoid doing this.
2. Explicitly list all of the potential subdomains: `ch-ua-platform-version=(self "https://abcd1234.akastat.io" "https://efgh5678.akastat.io" ...)`
    * We have 100s of potential subdomains, so maintaining this list would be a challenge
    * It would waste a lot of bandwidth
3. Use the JavaScript high-entropy API to get this data instead (in boomerang.js, the JS library that our customers use)
    * We are planning on doing this, but our customers are slow/hesitant to upgrade their boomerang.js versions so it may be years before they do so

Instead, we are hoping a proposal to allow for subdomains to match PP would be feasible and secure.

There was a bit of previous discussion on wildcards and subdomains here:

https://github.com/w3c/webappsec-permissions-policy/issues/129

And even a note from @clelland:

> CSP allows much richer wildcards than feature policy currently does -- If FP adopted some of the same parsing logic, then you could cut it down to

But as far as I can tell, wildcards/subdomains didn't make it into the final Client Hints spec.

I understand that "star" wildcards (e.g. `*.akstat.io`) may have security concerns, but we were hoping some sort of syntax that might allow for any subdomain of an origin to get data, maybe something like this amateur attempt at a syntax:

```
permissions-policy: ch-ua-platform-version=(self "subdomains:https://akstat.io" "https://c.go-mpulse.net") 
```

Anyways, we're just trying to do the right thing and not request these hints and blast them all over!

Would love your thoughts,

### PR DESCRIPTION
**Pull request by arichiv**
This change harnesses [Content Security Policy matching](https://www.w3.org/TR/CSP3/#matching-urls) to allow wildcards in the origins of Permissions Policy allowlists. The goal of this is to provide greater granularity in permissions to better balance the ease and risk of a wildcard permission against what might otherwise be a long list of similar, yet distinct origins.

The only wildcard allowed before was *, which matched all origins.

Wildcards will now be supported for:
* https://example.com:\* (ports, any port for example.com over https)
* https://\*.com (hosts, any https origin ending in .com)
* https: (scheme-only, any https origin)

Feature: https://chromestatus.com/feature/5101218029895680

closes #479

***
<a href="https://pr-preview.s3.amazonaws.com/w3c/webappsec-permissions-policy/pull/516.html" title="Last updated on Jun 9, 2023, 2:56 PM UTC (ae9d1b6)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/webappsec-permissions-policy/516/388f33c...ae9d1b6.html" title="Last updated on Jun 9, 2023, 2:56 PM UTC (ae9d1b6)">Diff</a>

### CODE REVIEW

File: index.bs
Comment (clelland) [inline index.bs:318]: I would still maintain that it is a set of origins, conceptually. It's all of the origins (potentially infinite) which match the various patterns used in the header or attribute. The match algorithm is essentially testing set inclusion.

Structurally, it is definitely a set of expressions, as you mention below. (This is all pretty pedantic, though)

File: index.bs
Comment (clelland) [inline index.bs:342]: The 0 here is irrelevant, right? It looks like that only comes into effect if there is a path component, which we won't pass in?

Comment (clelland) [reply index.bs:342]: Also, I'm not sure that we can pass `origin` in as the first parameter; the algorithm takes a URL, which an origin is not.

The second origin is interesting -- it should actually be the origin of the policy, which isn't part of the allowlist, since the algorithm handles items like 'self' here. However, we explicitly expand 'self' below when constructing allowlists. I wonder if this should be some null value instead, so that it cannot affect the match in any way?

File: index.bs
Comment (clelland) [inline index.bs:338]: We might want to revisit whether this step (above) is still necessary, and if so, point out that the behaviour diverges from CSP here.

Comment (arichiv) [reply index.bs:340]: Added a note

File: index.bs
Comment (arichiv) [reply index.bs:344]: We need to pass in the 0 since the algo asks for it though right?

Should I convert the origin to a string and then parse it as a URL? I assumed we could pass it here since it has a subset of the parts.

In the chrome source we are just passing the origin of the page twice https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/common/permissions_policy/origin_with_possible_wildcards.cc;drc=59515e08807151f06afcacfb2063732dd403ea05;l=116. I think that's the correct thing to do in this instance since we're doing all checks relative to the origin of the prospective page.

Comment (clelland) [reply index.bs:344]: > We need to pass in the 0 since the algo asks for it though right?

Yeah, I just wanted to make sure that it didn't actually matter -- you could pass in 7, or 100000000, and it wouldn't make a difference since there should be no path component.

> Should I convert the origin to a string and then parse it as a URL? I assumed we could pass it here since it has a subset of the parts.

`origin` can be opaque, in which case there would be no valid serialization, and we actually do need to see if it is same-origin with `item`. At least, that's the case currently. I think there may be no way for an allowlist to contain an opaque origin with this change.

That might be a more subtle change in the spec than I thought at first: Initially an allowlist would only contain concrete origins which were all resolved during parsing (unless the whole allowlist was `*`), but now it contains a list of patterns to match against, which cannot be made concrete.

> In the chrome source we are just passing the origin of the page twice https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/common/permissions_policy/origin_with_possible_wildcards.cc;drc=59515e08807151f06afcacfb2063732dd403ea05;l=116. I think that's the correct thing to do in this instance since we're doing all checks relative to the origin of the prospective page.

That algorithm, as I read it, takes \<absolute url of resource to check\>, \<pattern to check against\>, \<origin to resolve pattern against\>, \<redirect count\>. Because \<pattern\> could be `self`, which needs to resolve to the containing page's URL, that second origin should be the policy-owning-page's origin.

I think it doesn't matter right now, because we pre-resolve items like `self` during parsing, and don't store them in the allowlist, but if we need to change that, then the second origin will need to be that of the page with the policy.

Comment (arichiv) [reply index.bs:344]: Would it be enough to add a check that the origin wasn't opaque before calling this then?

File: index.bs
Comment (clelland) [inline index.bs:875]: This should be `src`, I believe.

File: index.bs
Comment (clelland) [inline index.bs:879]: Self should add `|container origin|` to the allowlist

**Review by clelland — COMMENTED** [review]
I think this is looking pretty good --  a couple of issues with `allow` parsing though.

**Comment by clelland** [comment]
Thanks -- I think this looks good -- can you file an issue on CSP to export the three terms that we use here? (That shouldn't block merging this, but it would be good cleanup for later)

The only thing that we might want to do to tighten it up a bit would be to give Allowlist two actual properties, so that we don't have to keep referring to "the origin representing `self` in allowlist", and could actually link to a definition there. That would probably require accounting for those properties to possibly be null. We'd be able to assume at that point that the "list" part of the allowlist is *just* patterns, as well.

(Not necessary right now; this is functionally equivalent, but if you wanted to do that before merging, let me know and I'll hold off)

**Comment by arichiv** [comment]
Filed https://github.com/w3c/webappsec-csp/issues/604

I think this can merge as is, I can circle back on that cleanup when the DFNs are exported by CSP.

**Review by clelland — APPROVED** [review]

**Pull request merged**
